### ISSUE TITLE
Dracut issues when booting with ZFS on root

### ISSUE THREAD

**Issue by PoisonFrog** [openzfs/zfs#13088]
First of all, I'm not technical or a sysadmin. I installed Arch Linux with ZFS on root (that's like the 5th time I'm doing it), but this time for some reason I get this at message at boot:

```
dracut-pre-mount[****]: cannot open ' ': name must begin with a letter
[FAILED] Failed to mount /sysroot
```
boot partition is encrypted with luks1
the disk where the zfs pool is stored is encrypted with luks2

initramfs opens the luks container and it imports the rpool but it doesn't mount the bootfs dataset (rpool/ROOT/default).

Anyone knows what this might be and how to fix it?

**Comment by szubersk** [comment]
@PoisonFrog you did not provide enough information to proceed with the investigation. Would you elaborate the installation process? Did you try `rd.break` kernel option to step into initrd scripts when they croak? Is there an easy way to reproduce this?
Maybe dracut fans in here would have a hunch what's going on.

**Comment by szubersk** [comment]
So actually there might be a bug here. `zfs:AUTO` didn't work correctly for initramfs too until I corrected it

**Comment by nabijaczleweli** [comment]
This is a bug, yes, we should handle all arguments via dracut-lib.sh `getarg`, but AFAICT from parse-zfs.sh.in, we're inheriting `$root` from... nowhere? so I'm not sure where it actually originates (if this were i-t then I'd assume cmdline->pid0->envp, but not under dracut!! dunno).

Upstream does this to obtain it (modules.d/99base/parse-root-opts.sh):
```sh
#!/bin/sh

# shellcheck disable=SC2034
root=$(getarg root=)

rflags="$(getarg rootflags=)"
getargbool 0 ro && rflags="${rflags},ro"
getargbool 0 rw && rflags="${rflags},rw"
rflags="${rflags#,}"

fstype="$(getarg rootfstype=)"
if [ -z "$fstype" ]; then
    fstype="auto"
fi
```
and (modules.d/95nfs/parse-nfsroot.sh):
```sh
 26 type getarg > /dev/null 2>&1 || . /lib/dracut-lib.sh
 27 . /lib/nfs-lib.sh
 28
 29 # This script is sourced, so root should be set. But let's be paranoid
 30 [ -z "$root" ] && root=$(getarg root=)
 31 [ -z "$nfsroot" ] && nfsroot=$(getarg nfsroot=)
 32
 33 [ -n "$netroot" ] && oldnetroot="$netroot"
```
so so should we.

**Comment by PoisonFrog** [comment]
> @PoisonFrog would you mind applying the topmost patch from #13093 (i.e.
> 
> ```diff
> diff --git a/contrib/dracut/90zfs/parse-zfs.sh.in b/contrib/dracut/90zfs/parse-zfs.sh.in
> index 724c5e2c6..2fa846f50 100755
> --- a/contrib/dracut/90zfs/parse-zfs.sh.in
> +++ b/contrib/dracut/90zfs/parse-zfs.sh.in
> @@ -16,6 +16,7 @@ else
>  fi
>  
>  wait_for_zfs=0
> +root="$(getarg root=)"
>  case "${root}" in
>  	""|zfs|zfs:)
>  		# We'll take root unset, root=zfs, or root=zfs:
> ```
> 
> ) to your `/usr/lib/dracut/modules.d/90zfs/parse-zfs.sh` and re-trying with the embedded dracut cmdline?

I can help, but again, I'm not technical. Should I just append this code to parse-zfs.sh file or is there anything else I need to do to apply the patch?

**Comment by PoisonFrog** [comment]
> Right, sorry. To apply this, add the `root="$(getarg root=)"` line just before the `case "${root}" in` line (after `wait_for_zfs=0`), such that you get
> 
> ```shell
> wait_for_zfs=0
> root="$(getarg root=)"
> case "${root}" in
> 	""|zfs|zfs:)
> ```
> 
> Then, to test, do the inverse of how you fixed it originally ‒ remove the `root=` stanza from your bootloader and re-add it to dracut. Thanks!

I added the `root="$(getarg root=)"` to the file right where you said it should be. Removed `root=zfs:AUTO` from /etc/default/grub and added all my kernel parameters to /etc/dracut.conf.d/cmdline.conf and the `root=zfs:AUTO` option. Then I ran `dracut --force`, `grub-mkconfig` and rebooted. I got the same error as before and could not boot.

**Comment by nabijaczleweli** [comment]
Damn; thanks for testing and leaving the steps for the dracut cmdline thing (which I haven't used before, hence why I haven't noticed the problem), I'll try to get a VM up to repro/fix this this week.

**Comment by PoisonFrog** [comment]
> Damn; thanks for testing and leaving the steps for the dracut cmdline thing (which I haven't used before, hence why I haven't noticed the problem), I'll try to get a VM up to repro/fix this this week.

This was not the first time I installed Arch Linux on ZFS like that and I didn't have this problem in my other installations. This was a first. I was also using dracut in all the other installations and to be honest I didn't even set the root=zfs:AUTO option anywhere and it was booting with no problems. I only had this problem with this bare metal install (the others were Virtual Box VMs). I tried dracut cmdline because I saw a similar issue where someone suggested the root=zfs:AUTO option and the person having the issue did it using dracut cmdline and it worked for him, but for me it didn't. Then I tried the root=zfs:AUTO option as a kernel parameter in /etc/default/grub and it worked for me.

You see. The root=zfs:AUTO [worked for someone else](https://github.com/openzfs/zfs/issues/13012#issuecomment-1022848111) in dracut cmdline so maybe this bug only happens when some conditions are met.
